### ISSUE TITLE
Infinite loop in g appication if pegs4 data too low in energy

### ISSUE THREAD

**Issue by dworogers** [nrc-cnrc/EGSnrc#634]
I accidentally used the wrong pegs4 file which only went up to 2 MeV when I was calculating for a 10 MeV photon beam.  The output .egslog just stops with no warning message but continues to use 100% of cpu. It needs a check.
I am using the feature-g-update branch version of g which worked fine when I switched to pegsless mode.

**Comment by ftessier** [comment]
Nice find! I have not found the solution yet (well, beyond running with the proper pegs4 file 😉 ). Looking at this through the debugger, the code appears to be locked up in the `electr` subroutine...

**Comment by ftessier** [comment]
A bit of investigating with `gdb` reveals that an electron ultimately gets stuck because `ustep .EQ. 0` in `IF ((ustep .EQ. 0 .OR. medium .EQ. 0))` on line 14174 (of the final `.f` Fortran code). So I know why the code hangs, but not yet why `ustep==0`. At any rate, gracefully trapping the error that the max energy is above the pegs limit at the outset would curb the issue; I would not try to trap this inside `electr`!

**Comment by dworogers** [comment]
Hi Frederic,
I agree that we want to catch it on input. The rz codes must catch it
somehow. I didn't have the patience to track how to find the highest energy
in a spectrum, but presumably it is done somewhere.

Thanks for looking at this.

As you say, using the right pegs4 data is the best solution, or even just
pegsless.

Dave

**Comment by ftessier** [comment]
@blakewalters can you suggest off-hand where we could add a check on the upper pegs energy, at the outset?

**Comment by mainegra** [comment]
@dworogers this would also happen if one uses an energy above the energy limits of the data regardless of whether one uses pegs4 or goes pegsless, right? I looked inside g and everything needed to perform the check is there. We just have to put the check, perhaps right after the call to the inputs routine. I could do it if @blakewalters is not already on it! 😜

**Comment by dworogers** [comment]
It seems to me we need to add a variable, perhaps to comin user (since
it is in subroutine ensrc already) which subroutine ensrc initializes
as the highest energy in a spectrum or EIN and then in main, when we
have access to UE,UP we can do the check.

**Comment by ftessier** [comment]
> @dworogers this would also happen if one uses an energy above the energy limits of the data regardless of whether one uses pegs4 or goes pegsless, right?

Interesting to note that using pegsless, at least there is some information printed showing that a particle is stuck. Not so with the pegs4 file. _Maybe just numerics throwing ustep on the negative side of `0` with pegsless?_

```
...
 993 Negative ustep = -0.24531E+00 dedx=  4.0360 ke= 57.9156 ir,irnew,irold =   1   1   1 x,y,z =-0.321E-06 0.420E-05 0.614E-01
 994 Negative ustep = -0.24531E+00 dedx=  4.0360 ke= 57.9156 ir,irnew,irold =   1   1   1 x,y,z =-0.321E-06 0.420E-05 0.614E-01
 995 Negative ustep = -0.24531E+00 dedx=  4.0360 ke= 57.9156 ir,irnew,irold =   1   1   1 x,y,z =-0.321E-06 0.420E-05 0.614E-01
 996 Negative ustep = -0.24531E+00 dedx=  4.0360 ke= 57.9156 ir,irnew,irold =   1   1   1 x,y,z =-0.321E-06 0.420E-05 0.614E-01
 997 Negative ustep = -0.24531E+00 dedx=  4.0360 ke= 57.9156 ir,irnew,irold =   1   1   1 x,y,z =-0.321E-06 0.420E-05 0.614E-01
 998 Negative ustep = -0.24531E+00 dedx=  4.0360 ke= 57.9156 ir,irnew,irold =   1   1   1 x,y,z =-0.321E-06 0.420E-05 0.614E-01
 999 Negative ustep = -0.24531E+00 dedx=  4.0360 ke= 57.9156 ir,irnew,irold =   1   1   1 x,y,z =-0.321E-06 0.420E-05 0.614E-01
1000 Negative ustep = -0.24531E+00 dedx=  4.0360 ke= 57.9156 ir,irnew,irold =   1   1   1 x,y,z =-0.321E-06 0.420E-05 0.614E-01
1001 Negative ustep = -0.24531E+00 dedx=  4.0360 ke= 57.9156 ir,irnew,irold =   1   1   1 x,y,z =-0.321E-06 0.420E-05 0.614E-01
```

**Comment by dworogers** [comment]
Hi,
I just checked and DOSRZnrc catches the problem and stops with

**************************************************
 FOR MEDIUM  1 INCIDENT ENERGY=      10.0 MeV
 IS GREATER THAN COVERED BY DATA FILE WHERE UP,UE=       2.0       2.5 MeV
     EXECUTION WILL TERMINATE AFTER COMPLETION OF INPUT FILE
 **************************************************
The example is just for a monoenergetic beam but when I ran a spectrum I
got the same message with the highest energy in the spectrum.

The call is just after call to hatch and just uses
    EKMAX=ENSRCD(NENSRC); "MAXIMUM KINETIC ENERGY IN THE SPECTRUM"
or EIN for a monoenergetic case.

Looks like that chunk of code would probably work in g.

**Comment by ftessier** [comment]
@mainegra want to handle this? Here is the [code chunk](https://github.com/nrc-cnrc/EGSnrc/blob/9a7165c1d76e21cd2072ef35620114e6527a447a/HEN_HOUSE/user_codes/dosrznrc/dosrznrc.mortran#L2627-L2651) in question.

**Comment by mainegra** [comment]
@dworogers  yes something like that would work in g. The slight complication comes from the fact that besides a spectrum and a unique energy users can define groups of more than one energy either individual values, linear or logarithmic grid. For that there is already a hook to the inputs routine that would/could return the maximum energy for all possible cases. Here is the current code inside subroutine inputs:
```
entry source_get_emax(emax);
"=========================="

IF( mono = 0 | mono = 2  | mono = 3 ) [
    IF( iqi = 0 ) [ emax = ei; ]
    ELSE          [ emax = ei-rm; ]
]
ELSE [
    emax = ensrcd(nensrc);
]
return;

```
which returns emax as you suggested. But is missing the cases of groups of energy. Which is easily expanded.

**Comment by mainegra** [comment]
> @mainegra want to handle this? Here is the [code chunk](https://github.com/nrc-cnrc/EGSnrc/blob/9a7165c1d76e21cd2072ef35620114e6527a447a/HEN_HOUSE/user_codes/dosrznrc/dosrznrc.mortran#L2627-L2651) in question.

@ftessier I can take care of it. Just wanted to make sure @blakewalters is not already working on it!

**Comment by rtownson** [comment]
@mainegra I believe Blake works on Mondays so you're probably safe.

**Comment by ftessier** [comment]
Can we think about trapping this at the top of the shower routine, as soon as we know the energy of the first (and only, `NP=1`) particle on the stack? That way we would not have to worry about it in every application...

**Comment by blakewalters** [comment]
@ftessier: Thumbs up!

**Comment by mainegra** [comment]
> Can we think about trapping this at the top of the shower routine, as soon as we know the energy of the first (and only, `NP=1`) particle on the stack? That way we would not have to worry about it in every application...

@ftessier then every app will have to provide Emax which will vary depending on the sources. At this point I wonder how different that is from putting the check in each application. The RZ codes have the srcrz.mortran common code, the C++ app via the base source class which will have to be implemented in each source .... which might already exist, ...

**Comment by ftessier** [comment]
Oh, I meant can we check  not Emax, but the actual energy of each new particle against UE or UP? But then I realized that this is perhaps wasteful. More importantly, for a low probability high energy line in the spectrum, the code might run for a long time before the unlikely high-energy source particle is sampled, and then the code would fail, which will be very confusing (unless we just discard the particle and issue a warning). So I am bak with @dworogers: put emax in a comin block and do the test once in the back end. Each app can update Emax as appropriate.

**Comment by rtownson** [comment]
I agree with checking Emax once, in the application for mortran codes. It's already done in egs++, in `egs_advanced_application.cpp`, line 698.

**Comment by dworogers** [comment]
Boy, did my one careless mistake lead to a flurry of emails! 
Chalk up one more to the idiot making things idiot-proof!

**Comment by mainegra** [comment]
> @mainegra: I wasn't working on this at the moment, but if you want me to do it, just let me know. I could also see about putting a similar check somewhere in the pegsless routines. I'd just have to check if the max. particle energy is available to those routines.

@blakewalters go ahead and implement the check in g! See the code above taken from inside subroutine inputs in g where emax is computed and add the cases for group of energies. Then in a similar fashion as @dworogers suggested, put the check after the HATCH call. Should be easy-peasy ... 😉

**Comment by mainegra** [comment]
> Boy, did my one careless mistake lead to a flurry of emails! Chalk up one more to the idiot making things idiot-proof!
> […](#)
> On Fri, 18 Sep 2020, Reid Townson wrote: I agree with checking Emax once, in the application for mortran codes. It's already done in egs++, in `egs_advanced_application.cpp`, line 698. -- You are receiving this because you were mentioned. Reply to this email directly or view it on GitHub: [#634 (comment)](https://github.com/nrc-cnrc/EGSnrc/issues/634#issuecomment-694986263)

@dworogers

**Comment by mainegra** [comment]
> Boy, did my one careless mistake lead to a flurry of emails! Chalk up one more to the idiot making things idiot-proof!
> […](#)
> On Fri, 18 Sep 2020, Reid Townson wrote: I agree with checking Emax once, in the application for mortran codes. It's already done in egs++, in `egs_advanced_application.cpp`, line 698. -- You are receiving this because you were mentioned. Reply to this email directly or view it on GitHub: [#634 (comment)](https://github.com/nrc-cnrc/EGSnrc/issues/634#issuecomment-694986263)

Just making it more robust @dworogers ! 😉

**Comment by blakewalters** [comment]
Okay, @mainegra, I'm on it...hopefully.

**Comment by ftessier** [comment]
Fixed in #635.
